standards/core/sync.md

```diff
@@ -7,126 +7,204 @@ contributors:
+in cases where the timestamp is the same.
+
+#### Range Bounds
+A _range_ MUST consists of 2 _IDs_, the first bound is
+inclusive the second bound exclusive.
+The first bound MUST be strictly smaller than the second one.
+
+#### Range Fingerprinting
+The _fingerprint_ of a range MUST be the XOR operation applied to
+the hash of all message _IDs_ included in that _range_.
+
+#### Range Type
+Every _range_ MUST have one of the following types; _fingerprint_, _skip_ or _item set_.
+
+- _Fingerprint_ type contains a _fingerprint_.
+- _Skip_ type contains nothing and is used to signal already processed _ranges_.
+- _Item set_ type contains message _Ids_ and a _resolved_ boolean.
+> _Item sets_ are an optimization, sending multiple _IDs_ instead of
+recursing further reduce the number of round-trips.
+
+#### Range Processing
+_Ranges_ have to be processed differently according to their types.
+
+- **Equal** _fingerprint ranges_ MUST become _skip ranges_.
+- **Unequal** _fingerprint ranges_ MUST be splitted into smaller ranges. The new type MAY be either _fingerprint_ or _item set_.
+- **Unresolved** _item set_ ranges MUST be checked for differences and marked resolved.
+- **Resolved** _item set_ ranges MUST be checked for differences and become skip ranges.
+- _Skip_ ranges MUST be merged with other consecutive _skip ranges_.
+
+In the case where only skip ranges remains, the synchronization is done.
+
+### Delta Encoding
+_Ranges_ and timestamps MUST be delta encoded as follows for efficient transmission.
+
+All _ranges_ to be transmitted MUST be ordered and only upper bounds used.
+> Inclusive lower bounds can be omitted because they are always
+the same as the exclusive upper bounds of the previous range or zero.
+
+To achieve this, it MAY be needed to add _skip ranges_.
+> For example, a _skip range_ can be added with
+an exclusive upper bound equal to the first range lower bound.
+This way the receiving peer knows to ignore the range from zero to the start of the sync time window.
+
+Every _ID_'s timestamps after the first MUST be noted as the difference from the previous one.
+If the timestamp is the same, zero MUST be used and the hash MUST be added.
+The added hash MUST be truncated up to and including the first differentiating byte.
+
+| Timestamp | Hash | Timestamp (encoded) | Hash (encoded) 
+| - | - | - | -
+| 1000 | 0x4a8a769a... | 1000 | -
+| 1002 | 0x351c5e86... | 2 | -
+| 1002 | 0x3560d9c4... | 0 | 0x3560
+| 1003 | 0xbeabef25... | 1 | -
+
+#### Varints
+All _varints_ MUST be little-endian base 128 variable length integers (LEB128) and minimally encoded.
+
+#### Payload encoding
+The wire level payload MUST be encoded as follow.
+> The & denote concatenation.
+
+1. _varint_ bytes of the delta encoded timestamp &
+2. if timestamp is zero, 1 byte for the hash bytes length & the hash bytes &
+3. 1 byte, the _range_ type &
+4. either
+    - 32 bytes _fingerprint_ &
+    - _varint_ bytes of the item set length & bytes of every items &
```

Comment:
Question, rather than a suggestion - the `&` makes it seem like these items will be concatenated, but if I understand correctly only one of the three listed are added. Perhaps "or" is a bit clearer after "either"?

```suggestion
    - 32 bytes _fingerprint_ or
    - _varint_ bytes of the item set length & bytes of every items or
```